app/master/build_request_handler.py

```diff
@@ -101,6 +101,11 @@ def _prepare_build_async(self, build, project_lock):
                                    log_msg='Build preparation loop is handling request for build {build_id}.')
             try:
                 build.prepare(self._subjob_calculator)
+                # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                if len(build.all_subjobs()) == 0:
+                    self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                    build.mark_started()
+                    build.perform_async_postbuild_tasks()
```

Comment:
Looking at this more -- I think I'd rather us keep `perform_async_postbuild_tasks` private. I think it was written intentionally to be executed on its own thread (started internally by the build). It feels a bit weird to execute it in one context on its own thread asynchronously and in another context synchronously, blocking other builds from being prepared.

I'm not really concerned about the *performance* of blocking other builds from preparing while doing postbuild tasks since this would only happen on builds with no subjobs, so no results, and the postbuild would be very fast. It's more of a conceptual thing.

The most direct way to solve that is probably to move the part in `Build` that starts the PostBuild thread into its own method, and then call that here instead.